Custom Actor Executors (#1952)
apple/swift-evolution: proposals/nnnn-custom-actor-executors.md

  * [Introduction](#introduction)
  * [Motivation](#motivation)
  * [Proposed solution](#proposed-solution)
  * [Detailed design](#detailed-design)
    + [A low-level design](#a-low-level-design)
    + [Executors](#executors)
    + [Serial Executors](#serial-executors)
    + [Jobs](#jobs)
    + [Actors with custom SerialExecutors](#actors-with-custom-serialexecutors)
    + [Executor assertions](#executor-assertions)
    + [Default Swift Runtime Executors](#default-swift-runtime-executors)
  * [Source compatibility](#source-compatibility)
  * [Effect on ABI stability](#effect-on-abi-stability)
  * [Effect on API resilience](#effect-on-api-resilience)
  * [Alternatives considered](#alternatives-considered)
  * [Future Directions](#future-directions)
    + [Overriding the MainActor executor](#overriding-the-mainactor-executor)
    + [Executor Switching](#executor-switching)
    + [Specifying Task executors](#specifying-task-executors)
    + [DelegateActor property](#delegateactor-property)

## Introduction

As Swift Concurrency continues to mature it is becoming increasingly important to offer adopters tighter control over where exactly asynchronous work is actually executed.

> This proposal introduces a basic mechanism for customizing actor  executors. By providing an instance of an executor, actors can influence "where" they will be executing any task they are running, while upholding the mutial excusion and actor isolation guaranteed by the actor model.

Reviewer: ```suggestion
This proposal introduces a basic mechanism for customizing actor executors. By providing an instance of an executor, actors can influence "where" they will be executing any task they are running, while upholding the mutual exclusion and actor isolation guaranteed by the actor model.
```

>  **Note:** This proposal defines only a set of APIs to customize actor executors, and other kinds of executor control is out of scope for this specific proposal.

## Motivation

> Swift's concurrency design is intentionally vague about the details of how code is actually run. Most code does not rely on specific properties of the execution environment, such as being run to a specific operating system thread, and instead needs only high-level semantic properties, expressed in terms of actor isolation, such as that no other code will be accessing certain variables concurrently. Maintaining flexibility about how work is scheduled onto threads allows Swift to avoid certain performance pitfalls by default.

Reviewer: ```suggestion
Swift's concurrency design is intentionally vague about the details of how code is actually run. Most code does not rely on specific properties of the execution environment, such as being run on a specific operating system thread, and instead needs only high-level semantic properties, expressed in terms of actor isolation, such as that no other code will be accessing certain variables concurrently. Maintaining flexibility about how work is scheduled onto threads allows Swift to avoid certain performance pitfalls by default.
```

This protocol has existed in Swift ever since the introduction of Swift Concurrency, however, in this proposal we revise its API to make use of the newly introduced move-only capabilities in the language. The existing `UnownedJob` API will be deprecated in favor of one accepting a move-only  `Job`. The `UnownedJob` type remains available (and equally unsafe), because today still some usage patterns are not supported by the initial revision of move-only types.

The concurrency runtime uses the `enqueue(_:)` method of an executor to schedule some work onto given executor.

```swift
/// A service that can execute jobs.
@available(SwiftStdlib 5.1, *)
public protocol Executor: AnyObject, Sendable {

  // This requirement is repeated here as a non-override so that we
  // get a redundant witness-table entry for it.  This allows us to
  // avoid drilling down to the base conformance just for the basic
  // work-scheduling operation.
  @available(SwiftStdlib 5.9, *)
  func enqueue(_ job: __owned Job)

  @available(SwiftStdlib 5.1, *)

>   @available(*, deprecated, message: "Implement the enqueue(Job) method instead")

Reviewer 2: Just as a heads up, this combination of availability attributes won't work correctly in that it is silently erasing the `introduced:` version for this API on platforms where an introduced version is needed. An `@available(*, ...)` attribute always takes precedence over everything else, so unfortunately you'll need to spell out the deprecation for each platform with explicit OS versions.

I have a couple bugs I'm tracking for this. One to warn you when you combine attributes like this since it will almost certainly behave unlike what you were expected. The other is to support availability macros (`SwiftStdlib 5.1` in this case) in long form `@available` attributes.

Reviewer 2: Actually, since `Executor` itself is introduced in `SwiftStdlib 5.1`, thankfully you don't need the `introduced` version on `enqueue` since it is inherited from the enclosing type and you can just omit `@available(SwiftStdlib 5.1, *)` instead of having the conflicting attributes.

Author: Ah great, thanks for spotting this!

This protocol has existed in Swift ever since the introduction of Swift Concurrency, however, in this proposal we revise its API to make use of the newly introduced move-only capabilities in the language. The existing `UnownedJob` API will be deprecated in favor of one accepting a move-only  `Job`. The `UnownedJob` type remains available (and equally unsafe), because today still some usage patterns are not supported by the initial revision of move-only types.

The concurrency runtime uses the `enqueue(_:)` method of an executor to schedule some work onto given executor.

```swift
/// A service that can execute jobs.
public protocol Executor: AnyObject, Sendable {

  // This requirement is repeated here as a non-override so that we
  // get a redundant witness-table entry for it.  This allows us to
  // avoid drilling down to the base conformance just for the basic
  // work-scheduling operation.
  func enqueue(_ job: consuming Job)

>   @available(SwiftStdlib 5.1, *)

Author: ```suggestion
```
